### PR TITLE
Remove monsters when switching to peaceful

### PR DESCRIPTION
**Pull request by Etmix**
Fix for the issue cuberite/cuberite#3622 .
This is the first time that I am working on a open source project and I just began reading into your codebase so any comment or help is appreciated.

### CODE REVIEW

**Comment by madmaxoft** [comment]
Unfortunately this is exactly the "fix" that doesn't really fix the problem - it removes the mobs from currently loaded chunks, but doesn't remove them from the other chunks that are loaded / generated later on.

I guess we can do this in two steps - have the Core plugin remove the mobs using your code, and then have another fix in the C++ code to remove mobs when loading / generating chunks.

**Comment by madmaxoft** [comment]
Oh wait, the plugin already blocks spawning. Now I'm just not sure if the HOOK_ENTITY_SPAWNING hook is being triggered for entities in a newly loaded chunk or not. I'll have to have a look in the code (and update the API docs).

Anyway, the code looks good except for some indent issues.

File: difficulties.lua [hunk 1]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
 	if (Split[2] == "peaceful") or (Split[2] == "0") or (Split[2] == "p") then
 		SetWorldDifficulty(Player:GetWorld(), 0)
 		SendMessage( Player, "World difficulty set to peaceful" )
+		
+		--Remove mobs which are not allowed in peaceful
+		Player:GetWorld():ForEachEntity(
+		function (a_Entity)
+			if not(a_Entity:IsMob()) then
+			return;
```
Comment (madmaxoft) [inline difficulties.lua]: Indent this line one tab more.

File: difficulties.lua [hunk 2]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
...
+		
+		--Remove mobs which are not allowed in peaceful
+		Player:GetWorld():ForEachEntity(
+		function (a_Entity)
+			if not(a_Entity:IsMob()) then
+			return;
+			end
+			local Monster = tolua.cast(a_Entity, "cMonster");
+			if (IsEntityBlockedInPeaceful[Monster:GetClass()]) then
+			Monster:Destroy();
```
Comment (madmaxoft) [inline difficulties.lua]: Indent this line one tab more.

**Comment by Etmix** [comment]
@madmaxoft I removed the semicolons.

@NiLSPACE You're right.

Thank you both for your comments :)

I created a new commit. Tell me if anything else needs to be changed.

File: difficulties.lua [hunk 3]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
 	if (Split[2] == "peaceful") or (Split[2] == "0") or (Split[2] == "p") then
 		SetWorldDifficulty(Player:GetWorld(), 0)
 		SendMessage( Player, "World difficulty set to peaceful" )
+		
+		--Remove mobs which are not allowed in peaceful
+		Player:GetWorld():ForEachEntity(
+		function (a_Entity)
+			if not(a_Entity:IsMob()) then
+			return
```
Comment (madmaxoft) [inline difficulties.lua]: Still missing an indent level here.

File: difficulties.lua [hunk 4]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
...
+		Player:GetWorld():ForEachEntity(
+		function (a_Entity)
+			if not(a_Entity:IsMob()) then
+			return
+			end
+			local Monster = tolua.cast(a_Entity, "cMonster")
+			if (Monster:GetMobFamily() == cMonster.mfHostile) then
+				Monster:Destroy()
+			end
+		end
```
Comment (madmaxoft) [inline difficulties.lua]: And the entire function (ending here) should have one more indent level as well.
```lua
...:ForEachEntity(
  function (...)
    -- function body
  end
)
```

**Comment by Etmix** [comment]
@madmaxoft Oh, I thought you added the indent levels when you mentioned them earlier.
Sorry for the inconvenience!

File: difficulties.lua [hunk 5]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
...
+			function (a_Entity)
+				if not(a_Entity:IsMob()) then
+					return
+				end
+				local Monster = tolua.cast(a_Entity, "cMonster")
+				if (Monster:GetMobFamily() == cMonster.mfHostile) then
+					Monster:Destroy()
+				end
+			end
+			)
```
Comment (madmaxoft) [inline difficulties.lua]: I know I'm being a nitpicker, but now this parenthesis has an extra indent level. It should be on the same level as the `ForEachEntity` call which it closes.

Of course I could fix that, too, but not in this PR (not allowed by GitHub), I'd need to create an extra commit  just for the fix after merging this PR, which would make the history messy.

Comment (Etmix) [inline difficulties.lua]: No problem. I appreciate that you take the time to correct me. 
I want to contribute to Cuberite more and want to do it right in order to learn as much as possible in the process.
I'll change it.

File: difficulties.lua [hunk 6]
```diff
@@ -24,6 +43,19 @@ function HandleDifficultyCommand ( Split, Player )
...
 		SendMessage( Player, "World difficulty set to peaceful" )
+		
+		--Remove mobs which are not allowed in peaceful
+		Player:GetWorld():ForEachEntity(
+			function (a_Entity)
+				if not(a_Entity:IsMob()) then
+					return
+				end
+				local Monster = tolua.cast(a_Entity, "cMonster")
+				if (Monster:GetMobFamily() == cMonster.mfHostile) then
```
Comment (madmaxoft) [inline difficulties.lua]: Could you try if this works without using `tolua.cast`? I believe the bindings should work in an intelligent way - if a `cEntity` class is sent to Lua, it should actually be the correct class already, so this should work directly:
```lua
-- (Keep the IsMob() check)
if (a_Entity:GetMobFamily() == cMonster.mfHostile) then
```

**Comment by madmaxoft** [comment]
The code is okay now, but I'd still like to know whether the `tolua.cast()` is needed (I don't think it is).

**Comment by NiLSPACE** [comment]
I tried to merge this, but Github gave a server error.

**Comment by madmaxoft** [comment]
Don't merge yet, I need answers! :)

**Comment by NiLSPACE** [comment]
You were saved by Github :-)

**Comment by Etmix** [comment]
@madmaxoft You are right. The cast is not needed. I removed it. :)

**Pull request merged**